apps/dashboard/src/components/workflow-row.tsx

```diff
@@ -0,0 +1,239 @@
+import { IEnvironment, WorkflowListResponseDto, WorkflowResponseDto } from '@novu/shared';
+import { RiAlertFill, RiDeleteBin2Line, RiGitPullRequestFill, RiPauseCircleLine, RiPulseFill } from 'react-icons/ri';
+import { Button } from '@/components/primitives/button';
+import {
+  DropdownMenu,
+  DropdownMenuContent,
+  DropdownMenuGroup,
+  DropdownMenuItem,
+  DropdownMenuSeparator,
+  DropdownMenuTrigger,
+} from '@/components/primitives/dropdown-menu';
+import { Link } from 'react-router-dom';
+import { TableCell, TableRow } from '@/components/primitives/table';
+import { useEnvironment } from '@/context/environment/hooks';
+import { WorkflowOriginEnum } from '@/utils/enums';
+import { buildRoute, LEGACY_ROUTES, ROUTES } from '@/utils/routes';
+import { usePromoteWorkflow } from '@/hooks/use-promote-workflow';
+import { useState } from 'react';
+import { Badge } from '@/components/primitives/badge';
+import { BadgeContent } from '@/components/primitives/badge';
+import { WorkflowStatus } from '@/components/workflow-status';
+import { WorkflowSteps } from '@/components/workflow-steps';
+import { WorkflowTags } from '@/components/workflow-tags';
+import { FaCode } from 'react-icons/fa6';
+import TruncatedText from '@/components/truncated-text';
+import {
+  Tooltip,
+  TooltipContent,
+  TooltipPortal,
+  TooltipProvider,
+  TooltipTrigger,
+} from '@/components/primitives/tooltip';
+import { RiMore2Fill, RiPlayCircleLine } from 'react-icons/ri';
+import { UseMutateAsyncFunction } from '@tanstack/react-query';
+import {
+  Dialog,
+  DialogClose,
+  DialogFooter,
+  DialogDescription,
+  DialogTitle,
+  DialogContent,
+  DialogOverlay,
+  DialogPortal,
+} from '@/components/primitives/dialog';
+
+type WorkflowRowProps = {
+  workflow: WorkflowListResponseDto;
+};
+
+export const WorkflowRow = ({ workflow }: WorkflowRowProps) => {
+  const { currentEnvironment } = useEnvironment();
+  const { safePromote, promote, isPromotable, tooltipContent } = usePromoteWorkflow(workflow);
+  const [showPromoteModal, setShowPromoteModal] = useState(false);
+
+  const handlePromote = async () => {
+    const { needsConfirmation } = await safePromote();
+
+    if (needsConfirmation) {
+      setShowPromoteModal(true);
+    } else {
+      await promote();
+    }
+  };
```

Comment:
All this is only relevant to the `PromoteWorkflowMenuItem` and could be encapsulated there, including the modal. Because there are cases when the modal is not needed at all, like the exact same cases when the menu item is `null`.

Reply:
I moved everything including the toast and modal to the `use-promote-workflow` hook and changed it to `.tsx`.

Although `usePromoteWorkflow` still needs to be used in `WorkflowRow` and not `PromoteWorkflowMenuItem` because the item is part of the dropdown menu and the menu including item disappears after clicking on item, so it would the state + modal etc.

Reply:
You can move that hook to be used only inside the `PromoteWorkflowMenuItem` with the modal included. As I mentioned before, you don't always need to render the modal.

Reply:
Please check the last comment - hook inside `PromoteWorkflowMenuItem` would lose state because when you chose/click some item from the menu, the menu automatically closes, therefore the state incl. modal needs to be at the top level of the menu. We can sync on that and I can showcase it directly.